torchaudio/datasets/cmudict.py

```diff
@@ -0,0 +1,105 @@
+from pathlib import Path
+from typing import Tuple, Union, List
+
+from torch.utils.data import Dataset
+from torchaudio.datasets.utils import download_url
+
+_CHECKSUMS = {
+    "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b":
+    "825f4ebd9183f2417df9f067a9cabe86",
+    "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b.symbols":
+    "385e490aabc71b48e772118e3d02923e",
+}
+
+
+class CMUDict(Dataset):
+    """Create a Dataset for CMU Pronouncing Dictionary (CMUDict).
+
+    Args:
+        root (str or Path): Path to the directory where the dataset is found or downloaded.
+        url (str, optional):
+            The URL to download the dictionary from.
+            (default: ``"http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b"``)
+        url_symbols (str, optional):
+            The URL to download the list of symbols from.
+            (default: ``"http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b.symbols"``)
+        download (bool, optional):
+            Whether to download the dataset if it is not found at root path. (default: ``False``).
+    """
+
+    def __init__(self,
+                 root: Union[str, Path],
+                 url: str = "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b",
+                 url_symbols: str = "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b.symbols",
+                 download: bool = False,
+                 exclude_punctuations: bool = True) -> None:
+
+        self.exclude_punctuations = exclude_punctuations
+
+        root = Path(root)
+        if not os.path.isdir(root):
+            os.mkdir(root)
+
+        if download:
+            if os.path.isdir(root):
+                checksum = _CHECKSUMS.get(url, None)
+                download_url(url, root, hash_value=checksum, hash_type="md5")
+                checksum = _CHECKSUMS.get(url_symbols, None)
+                download_url(url_symbols, root, hash_value=checksum, hash_type="md5")
+            else:
+                RuntimeError(f"The argument `root` must be a path to directory, "
+                             "but '{root}' is passed in instead.")
+
+        self._root_path = root
+        basename = os.path.basename(url)
+        basename_symbols = os.path.basename(url_symbols)
+
+        with open(os.path.join(self._root_path, basename_symbols), "r") as text:
+            self._symbols = text.readlines()
+
+        with open(os.path.join(self._root_path, basename), "r") as text:
+            self._dictionary = self._parse_dictionary(text.readlines())
+
+    def _parse_dictionary(self, lines: List[str]):
+        _alt_re = re.compile(r'\([0-9]+\)')
+        cmudict: List[Tuple[str, List[str]]] = list()
+        for line in lines:
+            if not line or line.startswith(';;;'):  # ignore comments
+                continue
+
+            word, phones = line.strip().split('  ')
+            if word[0] in string.punctuation:
```

Comment:
The content of `string.punctuation` is as follow;

```
$ python -c 'import string;print(string.punctuation)'
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
```

while the `word[0]` can take the following values.
The implementation looks correct. I feel safer if the mock dataset include at least one line example par the following unique first character.

```
$ grep -v ';;;' cmudict-0.7b | LC_CTYPE=C LANG=C cut -c-1 | sort -u
!
"
#
%
&
'
(
)
+
,
-
.
/
3
:
;
?
A
B
C
D
E
F
G
H
I
J
K
L
M
N
O
P
Q
R
S
T
U
V
W
X
Y
Z
{
}
```

Reply:
BTW `string.punctuation` is of `str` type. When performing `in` operation a lot, it is better to covert it to `set`.

Reply:
```
for char in `grep -v ';;;' cmudict-0.7b | LC_CTYPE=C LANG=C cut -c-1 | sort -u`; grep -m1 "^${char}" cmudict-0.7b;
!EXCLAMATION-POINT  EH2 K S K L AH0 M EY1 SH AH0 N P OY2 N T
"CLOSE-QUOTE  K L OW1 Z K W OW1 T
#HASH-MARK  HH AE1 M AA2 R K
%PERCENT  P ER0 S EH1 N T
&AMPERSAND  AE1 M P ER0 S AE2 N D
'ALLO  AA2 L OW1
(BEGIN-PARENS  B IH0 G IH1 N P ER0 EH1 N Z
)CLOSE-PAREN  K L OW1 Z P ER0 EH1 N
+PLUS  P L UH1 S
,COMMA  K AA1 M AH0
--DASH  D AE1 SH
;;; # CMUdict  --  Major Version: 0.07
/SLASH  S L AE1 SH
3-D  TH R IY1 D IY2
:COLON  K OW1 L AH0 N
;;; # CMUdict  --  Major Version: 0.07
?QUESTION-MARK  K W EH1 S CH AH0 N M AA1 R K
A  AH0
B  B IY1
C  S IY1
D  D IY1
E  IY1
F  EH1 F
G  JH IY1
H  EY1 CH
I  AY1
J  JH EY1
K  K EY1
L  EH1 L
M  EH1 M
N  EH1 N
O  OW1
P  P IY1
Q  K Y UW1
R  AA1 R
S  EH1 S
T  T IY1
U  Y UW1
V  V IY1
W  D AH1 B AH0 L Y UW0
X  EH1 K S
Y  W AY1
Z  Z IY1
{BRACE  B R EY1 S
}CLOSE-BRACE  K L OW1 Z B R EY1 S
```

Reply:
Now I've changed to using a [explicit list of punctuation](https://github.com/pytorch/audio/pull/1627/files#diff-50ee900e0897ac7528dccd9bdd36d5f0c7cb2d45310aefb46a3bf281f73d88c2R15) to filter out the punctuations. Could you also help me check if I missed any of the punctuations?
Thanks!